### PR TITLE
WorkManager sync service

### PR DESCRIPTION
**Pull request by deanrobertcook**
### To do: 
* remove Evernote Sync job and use WorkManager directly (separate PR)

### Tested so far:
* login and initial sync (this also tests the await method)
* sending messages and assets
* creating groups
* search
* all of the above on slow and flaky network
* ensure that failed message sending does not block sending of further messages
* sending large asset (or on flaky network) and minimising/swiping away application on Android >= 8 should not cause crash in background
* last read time
* Try sending a message in offline mode, try cancelling sending that message, try another one, then go back online 

### To Test:
* Performance on Android < 5.0 or on devices w/o GPS
* Upgrade with old sync jobs in database

### Current limitations:
* Sending asset blocks other messages into the same conversation until asset is sent (fix depends on assets refactoring)
#### APK
[Download build #12012](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12012/artifact/build/artifact/wire-dev-PR1857-12012.apk)
[Download build #12014](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12014/artifact/build/artifact/wire-dev-PR1857-12014.apk)
[Download build #12016](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12016/artifact/build/artifact/wire-dev-PR1857-12016.apk)
[Download build #12024](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12024/artifact/build/artifact/wire-dev-PR1857-12024.apk)
[Download build #12055](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12055/artifact/build/artifact/wire-dev-PR1857-12055.apk)
[Download build #12063](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12063/artifact/build/artifact/wire-dev-PR1857-12063.apk)
[Download build #12080](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12080/artifact/build/artifact/wire-dev-PR1857-12080.apk)
[Download build #12082](http://192.168.120.33:8080/job/Pull%20Request%20Builder/12082/artifact/build/artifact/wire-dev-PR1857-12082.apk)

### CODE REVIEW

File: app/src/main/scala/com/waz/background/WorkManagerSyncRequestService.scala [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    }
+  }
+
+  override def await(ids: Set[SyncId]): Future[Set[SyncResult]] =
+    Future.sequence(ids.map(await))
+
+  override def await(id: SyncId): Future[SyncResult] = {
+    implicit val logTag: LogTag = "WorkManager#await"
+    new LiveDataSignal(wm.getStatusByIdLiveData(UUID.fromString(id.str)))
+      .filter(_.getState.isFinished)
```
Comment (JPonte) [inline app/src/main/scala/com/waz/background/WorkManagerSyncRequestService.scala]: Filter and map is basically a collect no?

File: app/src/main/scala/com/waz/zclient/WireApplication.scala [hunk 2]
```diff
@@ -337,16 +346,16 @@ class WireApplication extends MultiDexApplication with WireContext with Injectab
...
 
-    ZMessaging.onCreate(this, backend, prefs, googleApi)
+    ZMessaging.onCreate(this, backend, prefs, googleApi, inject[WorkManagerSyncRequestService])
 
     inject[NotificationManagerWrapper]
     inject[ImageNotificationsController]
     inject[CallingNotificationsController]
 
-    //TODO [AN-4942] - is this early enough for app launch events?
+//    //TODO [AN-4942] - is this early enough for app launch events?
```
Comment (JPonte) [inline app/src/main/scala/com/waz/zclient/WireApplication.scala]: Not commented enough?

File: app/src/main/scala/com/waz/background/WorkManagerSyncRequestService.scala [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+                          req:        SyncRequest,
+                          priority:   Int            = Priority.Normal,
+                          dependsOn:  Seq[SyncId]    = Nil,
+                          forceRetry: Boolean        = false,
+                          delay:      FiniteDuration = Duration.Zero) = {
+
+    val work = new OneTimeWorkRequest.Builder(classOf[SyncJobWorker])
+      .setConstraints(
+        new Constraints.Builder()
+          .setRequiredNetworkType(NetworkType.CONNECTED)
```
[the remark this answers is not in the document]
Comment (deanrobertcook) [reply app/src/main/scala/com/waz/background/WorkManagerSyncRequestService.scala]: For now there's not a sync job that doesn't require the internet, so yes, but in the future we'll be introducing some non-internet-constrained jobs

**Comment by deanrobertcook** [comment]
build again

**Pull request merged**